File: scala/robot-simulator/src/main/scala/RobotSimulator.scala
object Bearing extends Enumeration {
  val North, East, South, West = Value
}

case class Robot(bearing: Bearing.Value, coordinates: (Int, Int)) {
  def turnLeft: Robot = {
    val bearings = Seq(Bearing.North, Bearing.West, Bearing.South, Bearing.East)
    copy(bearing = nextBearing(bearings))
  }

  def turnRight: Robot = {
    val bearings = Seq(Bearing.North, Bearing.East, Bearing.South, Bearing.West)
    copy(bearing = nextBearing(bearings))
  }

  def advance: Robot = {
    val (x, y) = coordinates
    val newCoordinates = bearing match {
      case Bearing.North => (x, y + 1)
      case Bearing.East => (x + 1, y)
      case Bearing.South => (x, y - 1)
      case Bearing.West => (x - 1, y)
    }
    copy(coordinates = newCoordinates)
  }

  def simulate(instructions: String): Robot =
    instructions.foldLeft(this)(_.simulate(_))

  private def nextBearing(bearings: Seq[Bearing.Value]): Bearing.Value =
    bearings.dropWhile(_ != bearing).drop(1).headOption.getOrElse(bearings.head)

  private def simulate(instruction: Char): Robot = instruction match {
    case 'A' => advance
    case 'L' => turnLeft
    case 'R' => turnRight
    case _ => throw new IllegalArgumentException(s"Unsupported instruction '$instruction'")
  }
}

File: fsharp/robot-simulator/RobotSimulator.fs
module RobotSimulator

type Direction =
    | North
    | East
    | South
    | West

type Position = int * int

let (+) ((x1, y1): Position) ((x2, y2): Position) = (x1 + x2, y1 + y2)

type Robot =
    { direction: Direction
      position: Position }

let create direction position =
    { direction = direction
      position = position }

let rightTurns =
    [ North; East; South; West; North ]
    |> List.pairwise
    |> Map.ofList

let leftTurns =
    [ North; West; South; East; North ]
    |> List.pairwise
    |> Map.ofList

let movement =
    function
    | North -> (0, 1)
    | East -> (1, 0)
    | South -> (0, -1)
    | West -> (-1, 0)

let step robot instruction =
    match instruction with
    | 'R' ->
        { robot with
              direction = rightTurns.[robot.direction] }
    | 'L' ->
        { robot with
              direction = leftTurns.[robot.direction] }
    | 'A' ->
        { robot with
              position = robot.position + movement robot.direction }
    | _ -> invalidOp "Valid instructions are 'L', 'R' and 'A'"

let move instructions robot = Seq.fold step robot instructions
